README.md
# OBJ3
OBJ3-2.10 - OBJ3 and TRIM compiler update to compile under GCL 2.6.12


ChangeLog
2003-09-29  Joseph R. Kiniry  <kiniry@acm.org>

	* Makefile:
	* RELEASE_NOTES.html:
	* TODO.html:
	* index.html:
	* source/README.html:
	* source/obj3/README.html:
	Final tweaks for the 2.09 release: details of lisp versions used, how
	to build OBJ3, zip usage for distribution archive, etc.

	* FAQ.html: Update for 2.09 release.

	* Makefile: Release building updated.

	* BUGS.html:
	* FAQ.html:
	* LICENSE.html:
	* README.html:
	* RELEASE_NOTES.html:
	* TODO.html:
	* docs/index.html:
	* docs/obj_bib.html:
	* index.html:
	* source/README.html:
	* source/obj3/README.html:
	Updates for 2.09 (CMU+GCL) release.

	* source/obj3/match/ac.lsp:
	* source/obj3/match/acz.lsp:
	* source/obj3/match/z.lsp:
	* source/obj3/modexp/mapping.lsp:
	* source/obj3/rew/substitution.lsp:
	* source/obj3/tools/obj_print.lsp:
	* source/obj3/top/misc.lsp:
	Final tweaks for CLISP compatibility.  While OBJ3 now builds with
	CLISP, it does not run properly.  Some debugging in lib/function.lsp
	is still necessary.  We'll leave this for the next release.

2003-09-26  Joseph R. Kiniry  <kiniry@acm.org>

	* source/obj3/basic/utilities.lsp: CLISP update.

	* Makefile: .PHONY added.

	* source/obj3/obj3.sh, source/obj3/obj3-load.sh:
	'in-package' modernization.  CMU and CLISP support updated.

	* source/obj3/obj3-lisp-dir.lsp:
	* source/obj3/obj3-files.lsp:
	'in-package' modernization.

	* source/obj3/obj3-compile.sh: Optimization and cleanup updates.

	* source/obj3/top/top.lsp:
	Eliminated get_pid/getpid in lieu of time-based generic implementation
	of temporary file name.  CLISP update, particularly wrt
	*standard-input* issues.

	* source/obj3/top/module_parse.lsp: CLISP *standard-input* support.

	* source/obj3/tools/tools.lsp: Compiled file suffix cleanup.

	* source/obj3/prelude/obj3sp.lsp:
	Cleanup of floating point types for all compilers.

	* source/obj3/lib/addr.lsp:
	* source/obj3/lib/bt.lsp:
	* source/obj3/lib/filecol.lsp:
	* source/obj3/lib/function.lsp:
	* source/obj3/lib/gctrace.lsp:
	* source/obj3/lib/hash.lsp:
	* source/obj3/lib/obless.lsp:
	* source/obj3/lib/sigs.lsp:
	* source/obj3/lib/type-of-safe.lsp:
	General cleanup, CMU support complete (but slow), and CLISP support
	improving.

	* source/obj3/Makefile:
	Added cleanup of compiled files from CMU on x86 and CLISP.

2003-09-23  Joseph R. Kiniry  <kiniry@acm.org>

	* source/obj3/obj3-quick-compile.sh: Chucking old files.

	* Makefile: 2.09/TRIM/CMU+GCL version.

	* source/trim/docs/trim-intro.ps:
	* source/trim/examples/FIB.eq:
	* source/trim/examples/FIB.trm:
	* source/trim/examples/INT.eq:
	* source/trim/src/Makefile:
	* source/trim/src/README:
	* source/trim/src/asm.hxx:
	* source/trim/src/asmlists.cxx:
	* source/trim/src/asmlists.hxx:
	* source/trim/src/basics.cxx:
	* source/trim/src/basics.hxx:
	* source/trim/src/config.h:
	* source/trim/src/config.hxx:
	* source/trim/src/const.hxx:
	* source/trim/src/eqcomp.cxx:
	* source/trim/src/eqcomp.hxx:
	* source/trim/src/eqgram.y:
	* source/trim/src/eqscan.l:
	* source/trim/src/fileio.c:
	* source/trim/src/fileio.h:
	* source/trim/src/mem.cxx:
	* source/trim/src/obstack.c:
	* source/trim/src/obstack.h:
	* source/trim/src/opt.cxx:
	* source/trim/src/opt.hxx:
	* source/trim/src/trim.cxx:
	* source/trim/src/trim.hxx:
	* source/trim/src/trim_instr.imp:
	* source/trim/src/trimasm.cxx:
	* source/trim/src/trimcc.c:
	* source/trim/src/trimcc.h:
	* source/trim/src/trimgram.y:
	* source/trim/src/trimmain.cxx:
	* source/trim/src/trimopt.cxx:
	* source/trim/src/trimscan.l:
	* source/trim/src/util.hxx:
	TRIM implementation from Lutz Hamel.

	* source/obj3/Makefile: Initial updates for 2.09 (GCL+CMU) release.

	* source/obj3/obj3-compile.sh, source/obj3/obj3-load.sh:
	Initial updates for 2.09 (GCL+CMU) release.  Compilation with
	optimization completes with GCL 2.5.3 and CMU 18e, but the resulting
	CMU image is terrifically slow.  GCL is the way to go right now until
	the CMU compilation process is debugged.

	* source/obj3/obj3-config.sh:
	* source/obj3/obj3-make.sh:
	* source/obj3/obj3.sh:
	Initial updates for 2.09 (GCL+CMU) release.

	* source/obj3/top/misc.lsp:
	* source/obj3/top/module_parse.lsp:
	* source/obj3/top/print.lsp:
	* source/obj3/top/top.lsp:
	* source/obj3/top/apply.lsp:
	Initial updates for 2.09 (GCL+CMU) release.  TRIM integration work
	completed by Lutz Hamel.

	* source/obj3/modexp/modexp.lsp:
	Initial updates for 2.09 (GCL+CMU) release.  Documentation cleanup.
	Eliminated make_symb and concat_string and inlined their definitions
	in defrepr to avoid a CMU compiler bug.  Also, made use of defrepr now
	for both CMU and GCL compilers.  Old defuns are still there, but
	commented out universally, but for functions that didn't have defrepr
	versions.

	* source/obj3/match/ac.lsp:
	Initial updates for 2.09 (GCL+CMU) release.  Documentation cleanup as
	well.

	* source/obj3/lib/filecol.lsp:
	* source/obj3/lib/gctrace.lsp:
	* source/obj3/lib/hash.lsp:
	* source/obj3/lib/obless.lsp:
	* source/obj3/lib/sigs.lsp:
	* source/obj3/lib/type-of-safe.lsp:
	Initial updates for 2.09 (GCL+CMU) release.  Foreign-function cleanup
	for dual CMU/GCL support.

	* source/obj3/lib/addr.lsp:
	* source/obj3/lib/alt.lsp:
	* source/obj3/lib/bt.lsp:
	* source/obj3/lib/function.lsp:
	* source/obj3/lib/macros.lsp:
	* source/obj3/lib/topo_sort.lsp:
	* source/obj3/match/match.lsp:
	* source/obj3/match/match_methods.lsp:
	* source/obj3/match/match_system.lsp:
	* source/obj3/match/state.lsp:
	* source/obj3/match/system.lsp:
	* source/obj3/match/theory.lsp:
	* source/obj3/match/theory_name.lsp:
	* source/obj3/match/theory_state.lsp:
	* source/obj3/match/z.lsp:
	* source/obj3/match/c.lsp:
	* source/obj3/match/cz.lsp:
	* source/obj3/match/empty.lsp:
	* source/obj3/match/environment.lsp:
	* source/obj3/match/global_state.lsp:
	* source/obj3/match/match_equation.lsp:
	* source/obj3/match/a.lsp:
	* source/obj3/match/acz.lsp:
	* source/obj3/match/az.lsp:
	* source/obj3/modexp/mapping.lsp:
	* source/obj3/modexp/modexp_eval2.lsp:

source/obj3/obj3-load.sh
#!/bin/bash

#
# OBJ3 2.06,2.08,2.09 Copyright (c) 2000-2003 Joseph Kiniry, Joseph Goguen
# OBJ3 2.05 Copyright (c) 2000 Sula Ma, Joseph Kiniry, Joseph Goguen
# OBJ3 2.04 Copyright (c) 1988,1991,1993 SRI International
# TRIM Copyright (c) 1994,2001 Lutz Hamel
# All Rights Reserved
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
# 
#   o Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# 
#   o Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
# 
#   o Neither name of the Joseph Kiniry, Joseph Goguen, Sula Ma, or SRI
#   International, nor the names of its contributors may be used to
#   endorse or promote products derived from this software without
#   specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SRI
# INTERNATIONAL OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
# OF THE POSSIBILITY OF SUCH DAMAGE.

#
# $Id: obj3-load.sh,v 206.1 2003/09/26 13:06:29 kiniry Exp $
#

#
# OBJ3 load
#

source obj3-config.sh

$lisp_compiler << \tio

(in-package :user)
(setq *print-case* :downcase)

(progn

  #+(or LUCID CMU) (load "impl.lsp")

  #+CMU (progn
          (setq *load-verbose* t)
          (defvar *load-print-stuff* t)
          (setq *block-compile-default* nil)
          (setq *compile-verbose* t)
          (setq *compile-print* t))

  #+GCL (setq si::*break-enable* t)

  #+GCL (progn 
          ;;please set your time zone
          (setq si::*default-time-zone* -1))

  (load "obj3-files.lsp" :verbose nil) (terpri)
        
  (print_char 60 "%") (terpri)
  (format t "Loading version: ~a~%" +version+)
  (print_char 60 "%") (terpri) 
        
  (dolist (file *obj3_files*)
    (load file)
    (print_char 60 "=") (terpri))

  (defvar obj3-load-time (get-time-string))
  (setq *obj3_files* nil)

  #+GCL (setq si::*break-enable* t)
  (obj3-init)

  (let (($$debug nil))
    (obj_input "prelude/obj3sp"))

  (defun #+GCL si:top-level #+(or LUCID CMU CLISP) obj3-top-level ()
         #+GCL (sig-setup)
         #+CMU (progn (lisp::set-auto-gc-trigger 2000000)
                      (gc-no-trace))
         (timer)
         (let ((res (catch *top-level-tag* (obj3) 'ok-exit)))
           (if (eq res 'ok-exit)
               #+GCL (bye) #+CMU (quit)
               (progn
                 (format t "OBJ3 loading error~%")
                 (terpri)))))

  #+GCL (progn (si:allocate 'cons 400)
               (si:allocate 'symbol 120)
               (si:allocate 'string 100)
               (si:allocate-relocatable-pages 300)
               (gbc t))
  #+LUCID (change-memory-management
           :growth-limit 1000
           :growth-rate 25
           :expand 25)
  #+CMU (gc)
  #+GCL (si::save-system "../../bin/obj3-gcl")
  #+LUCID (disksave "../../bin/obj3-lucid"
                    :full-gc t
                    :restart-function 'obj3-top-level)
  #+CMU (ext:save-lisp "../../bin/obj3-cmu"
                       :purify t
                       :init-function 'obj3-top-level
                       :print-herald nil
                       )
  #+CLISP (saveinitmem "../../bin/obj3-clisp"
                       :quiet t
                       :init-function 'obj3-top-level)
  #+(or GCL CLISP) (bye)
  #+CMU (quit)
  )
tio
